File: push-fe/src/shared/api/jobs.ts
import { api } from './client'
import { request } from './envelope'
import type { DataEnvelope, ListEnvelope, ListParams } from './envelope'
import { newIdempotencyKey } from '../lib/id'

export type JobPosting = {
  id: string
  revision: number
  company: string
  title: string
  sourceKind: 'URL' | 'TEXT' | 'DOM'
  sourceUrl: string | null
  sourceText: string
  requirements: string[]
  preferred: string[]
  keywords: string[]
  risks: string[]
  deadline: string | null
  language: string | null
  createdAt: string
  updatedAt: string
}

export type GapAnalysis = {
  id: string
  applicationId: string
  jobId: string
  jobRevision: number
  evidenceIds: string[]
  matched: { requirement: string; evidenceIds: string[] }[]
  missing: string[]
  preferredMissing: string[]
  risks: string[]
  fitScore: number | null
  method: 'RULE_BASED' | 'AI_ASSISTED'
  createdAt: string
}

export const listJobs = async (
  params: ListParams & { query?: string; archived?: boolean } = {},
): Promise<ListEnvelope<JobPosting>> =>
  request(() =>
    api.get('jobs', {
      searchParams: {
        ...(params.limit ? { limit: params.limit } : {}),
        ...(params.cursor ? { cursor: params.cursor } : {}),
        ...(params.query ? { query: params.query } : {}),
        ...(params.archived !== undefined
          ? { archived: String(params.archived) }
          : {}),
      },
    }),
  )

export const getJob = async (id: string): Promise<JobPosting> => {
  const env = await request<DataEnvelope<JobPosting>>(() =>
    api.get(`jobs/${id}`),
  )
  return env.data
}

export const createJob = async (body: {
  company: string
  title: string
  sourceKind: 'URL' | 'TEXT' | 'DOM'
  sourceUrl?: string
  sourceText: string
  requirements?: string[]
  preferred?: string[]
  deadline?: string
  language?: string
}): Promise<JobPosting> => {
  const env = await request<DataEnvelope<JobPosting>>(() =>
    api.post('jobs', {
      json: body,
      headers: { 'Idempotency-Key': newIdempotencyKey() },
    }),
  )
  return env.data
}

export const listJobAnalyses = async (
  jobId: string,
  params: ListParams = {},
): Promise<ListEnvelope<GapAnalysis>> =>
  request(() =>
    api.get(`jobs/${jobId}/analyses`, {
      searchParams: {
        ...(params.limit ? { limit: params.limit } : {}),
        ...(params.cursor ? { cursor: params.cursor } : {}),
      },
    }),
  )

File: push-be/app/presentation/routers/jobs.py
from __future__ import annotations

from fastapi import APIRouter, Request

from app.domain import entities as ent
from app.presentation import schemas as s
from app.presentation.deps import (
    bind_json, current_user, data, page_body, page_request, param_id,
    parse_deadline,
)

router = APIRouter()


def _deps(request: Request):
    return request.app.state.deps


def _ai(req: s.AiOptionsReq | None) -> ent.AiOptions | None:
    if req is None:
        return None
    return ent.AiOptions(provider=req.provider, model=req.model,
                         credential_mode=req.credential_mode, effort=req.effort)


def job_dto(j: ent.JobPosting) -> dict:
    return {"id": j.id, "revision": j.revision, "company": j.company,
            "title": j.title, "sourceKind": j.source_kind,
            "sourceUrl": j.source_url, "sourceText": j.source_text,
            "requirements": j.requirements, "preferred": j.preferred,
            "keywords": j.keywords, "risks": j.risks,
            "deadline": j.deadline.date() if j.deadline else None,
            "language": j.language, "createdAt": j.created_at,
            "updatedAt": j.updated_at}


@router.get("/jobs")
async def list_jobs(request: Request):
    d = _deps(request)
    page = page_request(request)
    archived = None
    if "archived" in request.query_params:
        archived = request.query_params["archived"] == "true"
    p = await d.jobs.list(current_user(request).id, archived,
                          request.query_params.get("query", ""), page)
    from app.jsonutil import to_jsonable
    from fastapi.responses import JSONResponse
    return JSONResponse(content={
        "data": [to_jsonable(job_dto(j)) for j in p.items],
        "page": {"nextCursor": p.next_cursor, "hasMore": p.has_more}},
        status_code=200)


@router.post("/jobs", status_code=201)
async def create_job(request: Request):
    d = _deps(request)
    req = await bind_json(request, s.CreateJobReq)
    deadline, _ = parse_deadline(req.deadline, "deadline" in req.model_fields_set)
    j = await d.jobs.create(current_user(request).id, req.company, req.title,
                            req.source_kind, req.source_url, req.source_text,
                            req.requirements, req.preferred, deadline,
                            req.language)
    return data(201, job_dto(j))


@router.get("/jobs/{id}")
async def get_job(id: str, request: Request):
    d = _deps(request)
    j = await d.jobs.get(current_user(request).id, param_id(id, "id"))
    return data(200, job_dto(j))


@router.patch("/jobs/{id}")
async def patch_job(id: str, request: Request):
    d = _deps(request)
    req = await bind_json(request, s.PatchJobReq)
    deadline, deadline_set = parse_deadline(
        req.deadline, "deadline" in req.model_fields_set)
    j = await d.jobs.patch(current_user(request).id, param_id(id, "id"),
                           req.expected_revision, req.company, req.title,
                           req.requirements, req.preferred, deadline,
                           deadline_set and deadline is None)
    return data(200, job_dto(j))


@router.post("/jobs/{id}/analyze", status_code=202)
async def analyze_job(id: str, request: Request):
    d = _deps(request)
    req = await bind_json(request, s.AnalyzeReq)
    op = await d.jobs.analyze(current_user(request).id, param_id(id, "id"),
                              req.application_id, req.expected_revision,
                              req.evidence_ids, _ai(req.ai))
    return data(202, op)


@router.get("/jobs/{id}/analyses")
async def list_analyses(id: str, request: Request):
    d = _deps(request)
    p = await d.jobs.list_analyses(current_user(request).id,
                                   param_id(id, "id"), page_request(request))
    return page_body(p)
